### PR TITLE
Add UART assisted tests

### PR DESCRIPTION
**Pull request by qiutongs**
Description
-----------

- Add UART assisted tests

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by qiutongs** [comment]
> we are not consistent on license, I thought we are using "Amazon FreeRTOS Common IO V1.0.0" for all files related to common-io, not "Amazon FreeRTOS V1.2.3"

I think you are right. Let me update them.

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+    def test_IotUartWriteAsync(self):
+        """
+        Test body of uart write async test. Return pass if expected message is received
+        :return: 'Pass' or 'Fail'
+        """
+        pi_result = None
+        dut_result = None
+        dut_res_pattern = re.compile(r"AFQP_AssistedIotUARTWriteAsync\W\sPASS")
+        pi_res_pattern = re.compile(r"\b(?:\\xaa){198}\b")
```
Comment (cmcgcc) [inline libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: why it is 198 here? Look like dut is sending out 200 bytes.

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        self.run_shell_script(
+            " ".join([self.shell_script, self._ip, self._login, self._pwd, "-c"])
+        )
+
+
+# unit test
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('-i', '--ip', nargs=1, default=['10.53.30.75'], help='ip address of rpi')
```
Comment (cmcgcc) [inline libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Could you take away the content in the default?

Comment (qiutongs) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Oh, sorry. Yes, I will remove

**Comment by qiutongs** [comment]
> I think the test coverage is still limited. The write read test could be separated as write sync and read sync, in case both are flipped. The read async, duplex sync/async should be covered as well. It is OK for this pr, but we should improve later.

Yeah, I agree. What in this PR is just a start for UART. Because assist tests is not the current priority. We will improve it later.

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py (hunk 1 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: https://github.com/aws/amazon-freertos/blob/2049152156e5ffc7d2d775da29b527487cc857b1/libraries/abstractions/common_io/test/test_iot_uart.c#L418 We lose one byte as the UART assisted tests use a '\n' byte to signal the write is done. In my PR I added a '\n' to the end of the buffer, but this change was lost when reviving that PR. We need to set cpBufferLarge[ testIotUART_BUFFER_LENGTH_LARGE  ] = '\n' and change this  to 199 bytes. I believe this was 198 as I was using '\r\n' but '\n' is sufficient and easier.

Comment (qiutongs) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: I will update it to 200

File: libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    def test_IotUartWriteAsync(self):
+        """
+        Test body of uart write async test. Return pass if expected message is received
+        :return: 'Pass' or 'Fail'
+        """
+        pi_result = None
+        dut_result = None
+        dut_res_pattern = re.compile(r"AFQP_AssistedIotUARTWriteAsync\W\sPASS")
+        # 200 corresponds to the config testIotUART_BUFFER_LENGTH_LARGE on the DUT
+        pi_res_pattern = re.compile(r"\b(?:\\xaa){200}\b")
```
Comment (lundinc2) [inline libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: This should be 199, as the last byte in the buffer is '\n', not 0xAA

Comment (qiutongs) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: The size of buffer is 201: testIotUART_BUFFER_LENGTH_LARGE

Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: The write is only 200 bytes. https://github.com/aws/amazon-freertos/blob/2049152156e5ffc7d2d775da29b527487cc857b1/libraries/abstractions/common_io/test/test_iot_uart.c#L428

Comment (qiutongs) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: Oh, You are right. Let me update the test to be:

```
 lWrite = iot_uart_write_async( xUartHandle, cpBufferLarge, testIotUART_BUFFER_LENGTH_LARGE + 1 );
```

Comment (cmcgcc) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: I do not see too much sense to modify c code just to keep 200 here. Let us just keep consistent with c code. It does not matter 199 or 200 here.

Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: I agree with @cmcgcc

Comment (qiutongs) [reply libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py]: ok

**Pull request merged**